src/class.cc

```diff
@@ -12,14 +12,65 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "class.h"
+
 namespace libwebp {
 
-// Let's implement stuff !
-// TODO: to create an animation:
-//       - create a WebPAnimEncoderOptions  config
-//       - set the quality to something in your search space
-//       - call WebPAnimEncoderNew, then fill the animation with WebPPicture
-//         using WebPAnimEncoderAdd
-//       - finalize with WebPAnimEncoderAssemble
+Thumbnailer::Thumbnailer() {
+  WebPAnimEncoderOptionsInit(&anim_config);
+  WebPConfigInit(&config);
+}
+
+Thumbnailer::~Thumbnailer() { WebPAnimEncoderDelete(enc); }
+
+bool Thumbnailer::AddFrame(const WebPPicture& pic, int timestamp_ms) {
+  bool ok = true;
+
+  // Verifies dimension of frames.
+  if (frames.empty()) {
+    enc = WebPAnimEncoderNew(pic.width, pic.height, &anim_config);
+    ok = (enc != nullptr);
+  } else {
+    if (pic.width != (frames[0].pic).width ||
+        pic.height != (frames[0].pic).height) {
+      ok = false;
+    }
+  }
+
+  FrameData new_frame = {pic, timestamp_ms};
+  frames.push_back(new_frame);
+
+  return ok;
+}
+
+bool Thumbnailer::GenerateAnimation(WebPData* const webp_data) {
+  bool ok = true;
+
+  // Rearranges frames.
+  std::sort(frames.begin(), frames.end(),
+            [&](FrameData A, FrameData B) -> bool {
+              return A.timestamp_ms < B.timestamp_ms;
+            });
+
+  // Fills the animation.
+  for (auto frame : frames) {
+    ok = ok && WebPAnimEncoderAdd(enc, &frame.pic, frame.timestamp_ms, &config);
+  }
+
+  // Adds last frame.
+  WebPAnimEncoderAdd(enc, NULL, frames.back().timestamp_ms, NULL);
+  WebPAnimEncoderAssemble(enc, webp_data);
+
+  // Sets loop count.
+  WebPMuxAnimParams new_params;
+  WebPMux* const mux = WebPMuxCreate(webp_data, 1);
```

Comment:
mux can be returned NULL in case of memory allocation failure. Catch this case.


I'd suggest catching errors right away went writing the initial code instead of leaving it as a TODO for later.
There's no real value in delaying the error-catching code. It *has* to be done, so better sooner than later, and be done with it.

Reply:
The best thing to do is encapsulate the deleter in a unique_ptr:
https://en.cppreference.com/w/cpp/memory/unique_ptr
std::unique_ptr<WebPMux*> mux(WebPMuxCreate(webp_data, 1), WebPMuxDelete);

This way, you can just check if it is null. If so, return kMemoryError right away.
Otherwise, you do not need to call WebPMuxDelete at the end, it will be called automatically.
And you do not need the if (ok) below. If you have an error, you can return right away.